far/editor.hpp

```diff
@@ -289,6 +289,8 @@ class Editor final: public SimpleScreenObject
 	enum class SearchReplaceDisposition;
 	SearchReplaceDisposition ShowSearchReplaceDialog(bool ReplaceMode);
 	void DoSearchReplace(SearchReplaceDisposition Disposition);
+	int CalculateSearchStartPosition(bool Continue, bool Backward, bool Regex) const noexcept;
```

Comment:
Let's not mark things `noexcept` unless necessary (e.g. it's a part of a contract or such a hot loop that it makes a difference).

Reply:
Sure, but why? I've always thought it a good idea to [[u]se `noexcept` when exiting a function because of a throw is impossible or unacceptable](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-noexcept).

The only reason not to blindly slap `noexcept` on a function which _cannot_ throw I am aware of is a public library, where breaking API changes are expensive, impractical or unacceptable, and the author cannot be sure future implementations can be kept `noexcept`. The key word here is _public library_.

Reply:
Exactly for the reason quoted: in principle, throw is neither impossible nor unacceptable in this case.
Indeed, it does not explicitly throw now because there is only some integer math inside, but it is an implementation detail than might change one day for any reason: we might call a function that throws or change that "assert" to "throw logic_error" etc.

If we blindly slap noexcept on every internal function that doesn't throw *today*, someone will have to go through the codebase, inspect the call tree and manually remove noexcept from all the affected functions every time the error handling logic changes somewhere, and there are more entertaining things in the world than that.

There are about 600 occurrences of noexcept in Far today and about 750 with plugins. Most of them were probably added by me, so I'm not rejecting the concept per se. However, noexcept is a commitment and should be treated with care to not cause more harm than good.

As a rule of thumb, if a function is longer than one line, if probably should not be noexcept.

Reply:
> If we blindly slap noexcept on every internal function that doesn't throw today, someone will have to go through the codebase, inspect the call tree and manually remove noexcept...

I see your point (and I kinda expected this argument), however:
Removing `noexcept` should happen when the function's implementation changes, not while fixing a crash. For internal functions, it's similar to renaming them or otherwise changing their signatures as things evolve. The `noexcept` spec is not stuck there forever. Anyway, I noticed your opinion and will follow the rules of the house of course (and if I forget, which may happen, please be patient with me).

What do you think of merging this PR? I have the next in the pipeline. :)

Reply:
P.S. My liking of `noexcept` stems from a real-world experience, when I traced enormous build times to huge code bloat caused by lack of `noexcept` on many relatively large but provably non-throwing functions. Even with forced heavy inlining, compiler could not prove everything and had to painstakingly generate megabytes of code to execute destructors as necessary. This issue is exacerbated by the (good) style of writing many small functions.

Reply:
> Removing `noexcept` should happen when the function's implementation changes

Indeed, but unfortunately it is not enforced by the compiler and thus it is too easy to get it wrong. We're only humans.
I wish `noexcept(auto)` was a thing, or better the default.

> What do you think of merging this PR?

LGTM, I can merge if you have nothing else to add to it (the discussion got quite long and I didn't follow everything).

Reply:
Please merge. I believe that my discussion with @HamRusTal is concluded (to mutual respectful disagreement), so there should not be problems here. I also tested the new behavior as best as I could. If @rohitab (or anybody else) finds any issues, I'll fix them separately.